### PR TITLE
Convert the cnv consensus output table to a segfile

### PR DESCRIPTION
**Pull request by jashapiro**
#### What scientific question is your analysis addressing?

How should various CNV calls be merged for downstream analysis?

#### What was your approach?

I wrote an R script to take the output from the copy number consensus analysis, and reduce it to a set of data that can be consumed as if it had come from cnvkit (more or less). Notably, as the copy number consensus may merge multiple nearby CNVs into a single event, this this requires some way of merging `seg.mean` values from multiple CNV calls. 

In order to merge `seg.mean` values, I am currently taking the approach of calculating the weighted mean of all CNVkit calls that comprise each consensus call. The code does include the possibility of various other schemes such as weighted medians with or without interpolation. 

While not used in this output, the script does also calculate a weighted median of copy number values (integers).

#### What GitHub issue does your pull request address?

#392

#### Which areas should receive a particularly close look?

Is the choice of weighted means to produce a single value reasonable?

I am using weighted median copy number from CNVkit where I can, and ControlFreeC values when CNVkit did not contribute to the consensus. Does that make sense?

#### Is there anything that you want to discuss further?

I do not think the discussions around #392 are complete: is this the format of output that we want to use, or should we include some of the other information in the final output file, such as alternative consensus seg.mean estimates?

#### What types of results are included (e.g., table, figure)?

A segfile table: `pbta-cnv-consensus.seg`

#### Reproducibility Checklist

- [x] The dependencies required to run the code in this pull request have been added to the project Dockerfile.
- [x] This analysis has been added to continuous integration.

#### Documentation Checklist

- [x] This analysis module has a `README` and it is up to date.
- [x] This analysis is recorded in the table in `analyses/README.md` and the entry is up to date.
- [x] The analytical code is documented and contains comments.

### CODE REVIEW

**Comment by jaclyn-taroni** [comment]
@jashapiro CI is failing with:

```
Complete log: /rocker-build/analyses/copy_number_consensus_call/.snakemake/log/2020-01-16T171857.497395.snakemake.log
Error in file.path(analysis_dir, opts$cnv_file) : 
  object 'analysis_dir' not found
Calls: <Anonymous> ... read_delimited -> source_name -> is.connection -> file.path
Execution halted
```

Full output is here: [build_2567_step_116_container_0.txt](https://github.com/AlexsLemonade/OpenPBTA-analysis/files/4073638/build_2567_step_116_container_0.txt)

**Comment by jaclyn-taroni** [comment]
@jashapiro I am unable to update this branch so it is up to date with the AlexsLemonade master branch. Can you allow edits from maintainers? If that is already enabled, then something else is going on and I'm not sure what it might be but I'm sure we could get to the bottom of it. Something similar happened on #439 so would love to figure it out if it's not that!

File: analyses/copy_number_consensus_call/results/pbta-cnv-consensus.seg [hunk 1]
```diff
@@ -0,0 +1,7070 @@
+ID	chrom	loc.start	loc.end	num.mark	seg.mean	copy.num
+BS_007JTNB8	chr11	771036	866778	NA	0.214821	3
```
[the remark this answers is not in the document]
Comment (jaclyn-taroni) [reply analyses/copy_number_consensus_call/results/pbta-cnv-consensus.seg]: Oh, does the `seg.mean` become `NA` when a call is the result of consensus between Manta and ControlFreeC but _not_ CNVkit?

File: analyses/copy_number_consensus_call/scripts/bed_to_segfile.R [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+# .seg output format from CNVkit is: 
+# ID	chrom	loc.start	loc.end	num.mark	seg.mean	copy.num
+
+out_table <- cnvs %>%
+  dplyr::select(ID = Biospecimen,
+                chrom = chrom,
+                loc.start = start, 
+                loc.end = end,
+                num.mark = num.mark,
+                seg.mean = segmean_wmean,
```
[the remark this answers is not in the document]
Comment (jashapiro) [reply analyses/copy_number_consensus_call/scripts/bed_to_segfile.R]: Yeah, I can do that. I'm not sure who will use it, but it makes for less dead code...

File: analyses/copy_number_consensus_call/results/pbta-cnv-consensus.seg (hunk 1 above)
[the remark this answers is not in the document]
Comment (jashapiro) [reply analyses/copy_number_consensus_call/results/pbta-cnv-consensus.seg]: That is correct. Only CNVkit uses seg.mean, so if there is no CNVkit call, we have nothing to work from. The num.mark column is also something potentially specific to CNVkit, but was not carried through in the consensus scripts. It seems to be optional to the .seg file format; I could just leave it off if nothing downstream depends on it. I thought preserving the column structure was worth doing for the first pass. 

As to the contents of num.mark: it is "bins" from CNVkit, so it doesn't really correspond to anything we can calculate easily, though I suppose we could fill in length or something like it.

[the remark this answers is not in the document]
Comment (jaclyn-taroni) [reply analyses/copy_number_consensus_call/results/pbta-cnv-consensus.seg]: What you've implemented sounds good to me in light of this additional information, thanks for the explainer!

File: analyses/copy_number_consensus_call/README.md [hunk 3]
```diff
@@ -35,6 +43,10 @@ Since there are 3 callers, there were 3 comparisons: `manta-cnvkit`, `manta-free
 
 11) **Sort and merge** the CNVs from the comparison pairs ,`manta-cnvkit` `manta-freec` `cnvkit-freec`, together into 1 file
 12) After every samples' consensus CNVs were called, **combine all merged files** from step 10 and output to `results/cnv_consensus.tsv`
+13) The `results/cnv_consensus.tsv` is translated into a `pbta-cnv-consensus.seg` file in the same format as `pbta-cnv-cnvkit.seg.gz`.
+When a consensus CNV contains from multiple source CNV segments, we take the mean of the CNVkit `seg.mean` values from the source segments, weighted by segment lenth.
```
Comment (jaclyn-taroni) [inline analyses/copy_number_consensus_call/README.md]:
```suggestion
When a consensus CNV contains from multiple source CNV segments, we take the mean of the CNVkit `seg.mean` values from the source segments, weighted by segment length.
```

**Pull request merged**